### PR TITLE
Store IP addresses in natural form

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

Instead of storing IP addresses as individually allocated strings, we store them in their natural byte form. Both IPv4 and IPv6 addresses fit into the 16 bytes variable we use for this (IPv4 addresses will only use the first four bytes, though).

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: datastructure.c [hunk 1]
```diff
@@ -83,16 +92,31 @@ int findOverTimeID(int overTimetimestamp)
 int findForwardID(const char * forward, bool count)
 {
 	int i, forwardID = -1;
+	int ret, proto = !(strstr(forward,":") != NULL) ? AF_INET : AF_INET6;
```
Comment (Mcat12) [inline datastructure.c]: Avoid a double negative (here and below)

File: datastructure.c [hunk 2]
```diff
@@ -112,7 +136,8 @@ int findForwardID(const char * forward, bool count)
 	else
 		forwarded[forwardID].count = 0;
 	// Save forward destination IP address
-	forwarded[forwardID].ip = strdup(forward);
+	// Store client IP
```
Comment (Mcat12) [inline datastructure.c]: Is this saving the client IP? Maybe delete this comment and use the above comment.

File: datastructure.c [hunk 3]
```diff
@@ -217,14 +254,34 @@ int findClientID(const char *client)
...
+	int ret, proto = clients[i].IPv4 ? AF_INET : AF_INET6;
+	if((ret = inet_pton(proto, ipaddr, clients[i].addr)) != 1)
+		logg("ERROR: inet_pton(%i, %s, %p) failed with %i", proto, ipaddr, clients[i].addr, ret);
+}
+
+char* getClientIP(int i)
+{
+	char *buffer = calloc(INET6_ADDRSTRLEN, sizeof(char));
+	int proto = clients[i].IPv4 ? AF_INET : AF_INET6;
+	inet_ntop(proto, clients[i].addr, buffer, INET6_ADDRSTRLEN);
```
Comment (Mcat12) [inline datastructure.c]: These get IP functions could check for IPv4 vs IPv6 before allocating.

File: dnsmasq_interface.c [hunk 4]
```diff
@@ -87,7 +87,7 @@ void FTL_new_query(unsigned int flags, char *name, struct all_addr *addr, char *
 	if(config.privacylevel >= PRIVACY_HIDE_DOMAINS_CLIENTS)
 	{
 		free(client);
-		client = strdup("0.0.0.0");
+		client = strdup("::");
```
Comment (Mcat12) [inline dnsmasq_interface.c]: Is this intended?

Comment (DL6ER) [reply dnsmasq_interface.c]: Yes, `0.0.0.0` is not really well-defined whereas `::` is officially the "unspecified address".

**Comment by DL6ER** [comment]
Closing this for now as sometimes spurious (wrong!) IPs appear. I will probably re-implement this in another form.

**Pull request closed without merging**